provider/throttle_utils.py

```diff
@@ -141,3 +152,600 @@ def change_throttle_group(self, image, group_id):
+        bps_write_max_length = attrs["bps-write-max-length"]
+        iops_read = attrs["iops-read"]
+        iops_read_max = attrs["iops-read-max"]
+        iops_read_max_length = attrs["iops-read-max-length"]
+        iops_total = attrs["iops-total"]
+        iops_total_max = attrs["iops-total-max"]
+        iops_total_max_length = attrs["iops-total-max-length"]
+        iops_write = attrs["iops-write"]
+        iops_write_max = attrs["iops-write-max"]
+        iops_write_max_length = attrs["iops-write-max-length"]
+
+        burst_read_iops = 0
+        burst_write_iops = 0
+        burst_total_iops = 0
+        normal_read_iops = 0
+        normal_write_iops = 0
+        normal_total_iops = 0
+
+        # reset expected result
+        self.set_throttle_expected(None, reset=1)
+
+        # count normal property
+        if iops_write != 0 or bps_write != 0:
+            bps = int(bps_write / iops_size)
+            iops = iops_write
+            if (iops >= bps != 0) or iops == 0:
+                normal_write_iops = bps
+            elif (bps >= iops != 0) or bps == 0:
+                normal_write_iops = iops
+            self.set_throttle_expected({"normal": {
+                "write": normal_write_iops}})
+
+        if iops_read != 0 or bps_read != 0:
+            bps = int(bps_read / iops_size)
+            iops = iops_read
+            if (iops >= bps != 0) or iops == 0:
+                normal_read_iops = bps
+            elif (bps >= iops != 0) or bps == 0:
+                normal_read_iops = iops
+            self.set_throttle_expected({"normal": {
+                "read": normal_read_iops}})
+
+        if iops_total != 0 or bps_total != 0:
+            bps = int(bps_total / iops_size)
+            iops = iops_total
+            if (iops >= bps != 0) or iops == 0:
+                normal_total_iops = bps
+            elif (bps >= iops != 0) or bps == 0:
+                normal_total_iops = iops
+            self.set_throttle_expected({"normal": {
+                "total": normal_total_iops}})
+
+        # count burst property
+        burst_time = 0
+        burst_empty_time = 0
+        if iops_write_max != 0 or bps_write_max != 0:
+            bps = int(bps_write_max * bps_write_max_length / iops_size)
+            iops = iops_write_max * iops_write_max_length
+            burst_write = 0
+            if (iops >= bps != 0) or iops == 0:
+                burst_write = bps
+                burst_write_iops = int(bps_write_max / iops_size)
+            elif (bps >= iops != 0) or bps == 0:
+                burst_write = iops
+                burst_write_iops = iops_write_max
+
+            empty_time = burst_write / normal_write_iops
+            full_time = burst_write / (burst_write_iops - normal_write_iops)
+            burst_empty_time = max(empty_time, burst_empty_time)
+            burst_time = max(full_time, burst_time)
+            self.set_throttle_expected({"burst": {"write": burst_write_iops}})
```

Comment:
This part, could we wrap them into a function as the above comment.
Thanks.

Reply:
please check above